Quantile control

MichaelXavier opened the pull request:
@ozataman I'd like a review of the approach I took here. This adds a config record to AggProcess, so that it isn't just a type alias for the aggregation function. This config record has a field with a function mapping from metric name to a set of quantiles and some combinators for common formulations of that mapping. Now that the user can limit quantiles per-worker, the cloudwatch worker can actually take any quantiles that make it through and publish them as sub metrics. My initial tests of this appear to be working correctly.

I explain in the commit log but I decided to use a function because not all setups can enumerate all of the metric names to just make a simple `Map MetricName (Set Quantile)`. Some metric backends may bake some dimension info into the metric name, and thus they would want to match on names with a function.

I'm currently dogfooding this internally and it appears to work. If you prefer, you can forgo merging this until we have mileage in production, but let me know if you approve of the direction.

MichaelXavier committed:
Spike on configuring workers to control quantiles

We'll see how this plays out, but the idea here is to allow each
worker to decide which quantiles get published. This is especially
critical for cloudwatch, which does not allow you to submit quantile
metrics directly. So that means we need to create sub-metrics for each
quantile and because of how heavy that is, it is really important to
allow the user to curtail the quantiles that get published.

I tried to keep things simple and just take a map from metric name to
quantile set, but I think this is not appropriate for other
backends. Other backends (like graphite IIRC), encode information into
their metric names rather than dimensions. This means that there's a
runtime component to the metric name. For those backends, if they
wanted to use this feature and not just use standardQuantiles, they
would need a function to match the shape of a metric key, rather than
just specify it fully with a map. I tried to provide some combinators
to make this simpler.

MichaelXavier committed:
Move Quantile to Instrument.Types

MichaelXavier committed:
Publish quantile submetrics in cloudwatch instrument

ozataman commented:
I think the overall direction makes sense. The couple of helpers make the function approach very close to the fixed map, so that doesn't seem to be an issue. 

This also implicitly adds the flexibility to decide which quantiles should be emitted for all metrics even for simpler cases like the CSV backend. (E.g. just emit p90 and p99 for all metrics, not the rest of the full scala.)

I'm assuming this will 100% not change any existing metrics in our current prod, but will only add the new quantiles, right?

MichaelXavier commented:
> I'm assuming this will 100% not change any existing metrics in our current prod, but will only add the new quantiles, right?

To continue behavior exactly as it exists today, you'd just use `standardQuantiles` for your backend (for graphite and CSV). For CloudWatch, to replicate what you'd have today, you'd use `noQuantiles` because the version on `master` ignores quantiles entirely. Adding quantiles for a CloudWatch backend with this patch will leave metrics as they're currently reported today, but it will also emit a variant of that metric per-quantile.

Since it sounds like this is okay to move forward, I'll keep this in a branch, wait til there's prod mileage on this and then merge. I'm pretty confident that will go well as I've already locally tested some metrics with the CloudWatch backend and selective quantiles and the graphs looks as I'd expected.

MichaelXavier committed:
Consistently use timer prefix in submitTime

This feature hasn't gone to master yet but since submitTime was split
out from timeI, it should be responsible for prepending the "time."
prefix, rather than timeI, that way both are guaranteed to produce the
correct name.

MichaelXavier committed:
Make timerMetricName idempotent

MichaelXavier committed:
Work around timer prefix more reliably

Starting to think it would be better to just pull the bandage off and
drop time prefixing altogether, but I'll leave that stone
unturned. This ensures that the metric name will be checked for
dimensions both with and without the time prefix.
